lale/lib/aif360/orbis.py

```diff
@@ -46,186 +48,20 @@
-    ndi = odi + bias_repair_level * (1 - odi)
-    # we have two equations, one each for nci and ndi
-    #    nci == (n00 + n10) / (n01 + n11)
-    #    ndi == (n01 / (n00 + n01)) / (n11 / (n10 + n11))
-    # without loss of generality, assume oci <= 1 and odi <= 1
-    assert oci <= 1 and odi <= 1, _sizes_to_string(osizes, "o")
-    # that means we do not need to upsample o11
-    # we will set n11 == o11, leaving three unknowns: n00, n01, n10
-    # two equations admit multiple solutions for the three unknowns
-    # algorithm to pick the solution that minimizes the amount of oversampling:
-    # - loop over candidate values for n00 in ascending order
-    # - given n00, solve the equations to also pick values for n01 and n10
-    # - terminate when all group sizes >= their original size
-
-    invalid = {"00": -1, "01": -1, "10": -1, "11": -1}
-
-    def solve_for_n01_n10_given_00_11(n00, n11):
-        # rewriting the nci equation:
-        #    n01 == n00 / nci + n10 / nci - n11
-        # rewriting the ndi equation:
-        #    ndi * n11 * n00 + ndi * n11 * n01 == n01 * n10 + n01 * n11
-        # substituting n01 into the ndi equation:
-        #    ndi*n11*n00 + ndi*n11*(n00/nci + n10/nci - n11)
-        #    == n10*(n00/nci + n10/nci - n11) + n11*(n00/nci + n10/nci - n11)
-        # rewriting this to the standard form of a quadratic equation for n10:
-        #    n10*n10
-        #      + n10*(n00 + n11 - n11*nci - ndi*n11)
-        #      + (n11*n00+ndi*n11*n11*nci-n11*n11*nci-ndi*n11*n00*nci-ndi*n11*n00)
-        #    == 0
-        # assigning variables so the above is n10*n10 + n10 * b + c == 0:
-        b = n00 + n11 - n11 * nci - ndi * n11
-        c = (
-            n11 * n00
-            + ndi * n11 * n11 * nci
-            - n11 * n11 * nci
-            - ndi * n11 * n00 * nci
-            - ndi * n11 * n00
-        )
-        # the square root of a negative number is imaginary
-        if b * b - 4 * c < 0:
-            return invalid
-        # quadratic equations have two solutions
-        n10_plus = (-b + (b * b - 4 * c) ** 0.5) / 2
-        n01_plus = n00 / nci + n10_plus / nci - n11
-        valid_plus = round(n01_plus) >= o01 and round(n10_plus) >= o10
-        n10_minus = (-b - (b * b - 4 * c) ** 0.5) / 2
-        n01_minus = n00 / nci + n10_minus / nci - n11
-        valid_minus = round(n01_minus) >= o01 and round(n10_minus) >= o10
-        if valid_plus and valid_minus:  # pick solution minimizing n01 + n10
-            if n01_plus + n10_plus < n01_minus + n01_minus:
-                n01, n10 = n01_plus, n10_plus
-            else:
-                n01, n10 = n01_minus, n10_minus
-        elif valid_plus:
-            n01, n10 = n01_plus, n10_plus
-        elif valid_minus:
-            n01, n10 = n01_minus, n10_minus
-        else:
-            return invalid
-        _assert_almost_equal(nci, _class_imbalance(n00, n01, n10, n11))
-        _assert_almost_equal(ndi, _disparate_impact(n00, n01, n10, n11))
```

Comment:
I wonder if we can have an assertion similar to this one somewhere. Basically, this checks that the solution actually achieves the goal class imbalance and disparate impact level. Maybe this could be in the tests?

Reply:
Implemented some helper functions in `_mystic_util.py` to make something like this possible.